// @applitools/visual-grid-client/src/sdk/EyesWrapper.js
'use strict';
const {EyesBase, NullRegionProvider} = require('@applitools/eyes-sdk-core');

const VERSION = require('../../package.json').version;

class EyesWrapper extends EyesBase {
  constructor({apiKey, logHandler} = {}) {
    super();
    apiKey && this.setApiKey(apiKey);
    logHandler && this.setLogHandler(logHandler);
  }

  async open(appName, testName, viewportSize) {
    await super.openBase(appName, testName);

    if (viewportSize) {
      this.setViewportSize(viewportSize);
    }
  }

  async ensureAborted() {
    if (!this.getRunningSession()) {
      this._configuration.mergeConfig(this.getAssumedConfiguration());
      await this._ensureRunningSession();
    }
    await this.abort();
  }

  setAssumedConfiguration(configuration) {
    this._assumedConfiguration = configuration;
  }

  getAssumedConfiguration() {
    return this._assumedConfiguration;
  }

  /** @override */
  getBaseAgentId() {
    return this.agentId || `visual-grid-client/${VERSION}`;
  }

  setBaseAgentId(agentId) {
    this.agentId = agentId;
  }

  /**
   * Get the AUT session id.
   *
   * @return {Promise<?String>}
   */
  async getAUTSessionId() {
    return; // TODO is this good?
  }

  /**
   * Get a RenderingInfo from eyes server
   *
   * @return {Promise<RenderingInfo>}
   */
  getRenderInfo() {
    return this._serverConnector.renderInfo();
  }

  setRenderingInfo(renderingInfo) {
    this._serverConnector.setRenderingInfo(renderingInfo);
  }

  /**
   * Create a screenshot of a page on RenderingGrid server
   *
   * @param {RenderRequest[]} renderRequests - The requests to be sent to the rendering grid
   * @return {Promise<String[]>} - The results of the render
   */
  renderBatch(renderRequests) {
    renderRequests.forEach(rr => rr.setAgentId(this.getBaseAgentId()));
    return this._serverConnector.render(renderRequests);
  }

  putResource(runningRender, resource) {
    return this._serverConnector.renderPutResource(runningRender, resource);
  }

  getRenderStatus(renderId) {
    return this._serverConnector.renderStatusById(renderId);
  }

  checkWindow({screenshotUrl, tag, domUrl, checkSettings, imageLocation, source}) {
    const regionProvider = new NullRegionProvider();
    this.screenshotUrl = screenshotUrl;
    this.domUrl = domUrl;
    this.imageLocation = imageLocation;
    return this.checkWindowBase(regionProvider, tag, false, checkSettings, source);
  }

  async getScreenshot() {
    return;
  }

  async getScreenshotUrl() {
    return this.screenshotUrl;
  }

  async getInferredEnvironment() {
    return this.inferredEnvironment;
  }

  setInferredEnvironment(value) {
    this.inferredEnvironment = value;
  }

  async setViewportSize(viewportSize) {
    this._configuration.setViewportSize(viewportSize);
    this._viewportSizeHandler.set(this._configuration.getViewportSize());
  }

  async getTitle() {
    return 'some title'; // TODO what should this be? is it connected with the tag in `checkWindow` somehow?
  }

  async getDomUrl() {
    return this.domUrl;
  }

  async getImageLocation() {
    return this.imageLocation;
  }
}

module.exports = EyesWrapper;
